File: Design Pattern/Structural/Composite/organization.cpp
#include <algorithm>
#include <iostream>
#include <string>
#include <vector>

using namespace std;

class OrgComponent {
public:
  virtual int getSalary() = 0;
  virtual int getHeadcount() = 0;
  virtual void printHeirarchy(string indent) = 0;
  virtual ~OrgComponent() {}
};

class Employee : public OrgComponent {
private:
  string name, title;
  int salary;

public:
  Employee(string name, string title, int salary)
      : name(name), title(title), salary(salary) {}

  int getSalary() override { return salary; }

  int getHeadcount() override { return 1; }

  void printHeirarchy(string indent) override {
    cout << indent << "- " << name << " (" << title << ", $" << salary << ")"
         << endl;
  }
};

class Manager : public OrgComponent {
private:
  string name, title;
  int salary;
  vector<OrgComponent *> members;

public:
  Manager(string name, string title, int salary)
      : name(name), title(title), salary(salary) {}

  void addMember(OrgComponent *member) { members.push_back(member); }

  void removeMember(OrgComponent *member) {
    members.erase(remove(members.begin(), members.end(), member),
                  members.end());
  }

  int getSalary() override {
    int total = salary;
    for (OrgComponent *member : members)
      total += member->getSalary();
    return total;
  }

  int getHeadcount() override {
    int count = 1;
    for (OrgComponent *member : members)
      count += member->getHeadcount();
    return count;
  }

  void printHeirarchy(string indent) override {
    cout << indent << "+ " << name << " (" << title << ", $" << salary << ")"
         << endl;
    for (OrgComponent *member : members)
      member->printHeirarchy(indent + "  ");
  }
};

int main() {
  Employee dev1("Alice", "Senior Engineer", 120000);
  Employee dev2("Bob", "Engineer", 95000);
  Employee dev3("Charlie", "Engineer", 90000);
  Employee designer("Diana", "Designer", 100000);

  Manager techLead("Eve", "Tech Lead", 140000);
  techLead.addMember(&dev1);
  techLead.addMember(&dev2);

  Manager vpEng("Frank", "VP Engineering", 200000);
  vpEng.addMember(&techLead);
  vpEng.addMember(&dev3);

  Manager vpProduct("Grace", "VP Product", 190000);
  vpProduct.addMember(&designer);

  Manager ceo("Hank", "CEO", 300000);
  ceo.addMember(&vpEng);
  ceo.addMember(&vpProduct);

  cout << "---- Organization Chart ----" << endl;
  ceo.printHeirarchy("");

  cout << "\nTotal Payroll: $" << ceo.getSalary() << endl;
  cout << "Total Headcount: " << ceo.getHeadcount() << endl;
  cout << "\nEngineering Payroll: $" << vpEng.getSalary() << endl;
  cout << "Engineering Headcount: " << vpEng.getHeadcount() << endl;

  return 0;
}

File: Design Pattern/Structural/Composite/organization.ts
interface OrgComponent {
  getSalary(): number;
  getHeadcount(): number;
  printHierarchy(indent: string): void;
}

class Employee implements OrgComponent {
  private readonly name: string;
  private readonly title: string;
  private readonly salary: number;

  constructor(name: string, title: string, salary: number) {
    this.name = name;
    this.title = title;
    this.salary = salary;
  }

  getSalary(): number {
    return this.salary;
  }

  getHeadcount(): number {
    return 1;
  }

  printHierarchy(indent: string): void {
    console.log(`${indent}- ${this.name} (${this.title}, $${this.salary})`);
  }
}

class Manager implements OrgComponent {
  private readonly name: string;
  private readonly title: string;
  private readonly salary: number;
  private readonly members: OrgComponent[] = [];

  constructor(name: string, title: string, salary: number) {
    this.name = name;
    this.title = title;
    this.salary = salary;
  }

  addMember(member: OrgComponent): void {
    this.members.push(member);
  }

  removeMember(member: OrgComponent): void {
    const index = this.members.indexOf(member);
    if (index !== -1) this.members.splice(index, 1);
  }

  getSalary(): number {
    let total = this.salary;
    for (const member of this.members) total += member.getSalary();
    return total;
  }

  getHeadcount(): number {
    let count = 1;
    for (const member of this.members) count += member.getHeadcount();
    return count;
  }

  printHierarchy(indent: string): void {
    console.log(`${indent}+ ${this.name} (${this.title}, $${this.salary})`);
    for (const member of this.members) member.printHierarchy(indent + "  ");
  }
}

const dev1 = new Employee("Alice", "Senior Engineer", 120000);
const dev2 = new Employee("Bob", "Engineer", 95000);
const dev3 = new Employee("Charlie", "Engineer", 90000);
const designer = new Employee("Diana", "Designer", 100000);

const techLead = new Manager("Eve", "Tech Lead", 140000);
techLead.addMember(dev1);
techLead.addMember(dev2);

const vpEng = new Manager("Frank", "VP Engineering", 200000);
vpEng.addMember(techLead);
vpEng.addMember(dev3);

const vpProduct = new Manager("Grace", "VP Product", 190000);
vpProduct.addMember(designer);

const ceo = new Manager("Hank", "CEO", 300000);
ceo.addMember(vpEng);
ceo.addMember(vpProduct);

console.log("---- Organization Chart ----");
ceo.printHierarchy("");

console.log("\nTotal Payroll: $" + ceo.getSalary());
console.log("Total Headcount: " + ceo.getHeadcount());
console.log("\nEngineering Payroll: $" + vpEng.getSalary());
console.log("Engineering Headcount: " + vpEng.getHeadcount());
